[FEATURE] ZEP - dynamic pandas asset schema definitions

Kilo59 opened the pull request:
The goal in this PR is to be able to dynamically create `pandas` `DataAsset` types by inspecting the function signatures of the `pandas.io` read methods (eg `read_csv`, `read_excel`, `read_json` etc.). `20` methods in total.

The names and type annotations from these function signatures will be used to define fields on the corresponding `DataAsset` example: (`CSVAsset`, `ExcelAsset` etc.).

~There are some arguments and types that we will ignore during this initial effort due to difficulties representing the types during serialization.~ Example: `Callable`. We could choose to ignore these only during serialization but then we could get into an awkward state where our `DataAsset` exhibited different behavior the first time we created in code, than it does after we round trip it from config.

If a type has no annotation or otherwise cannot be serialized, in this initial effort we will either give it an `Any` type, or a more specific primitive type. Example: `dtype` -> `dict`

## Changes proposed in this pull request:
- Generate pandas `DataAsset` types dynamically by inspecting the `pandas.io` read methods
- change existing `CSVAsset` -> `_DataFrameAsset` which all dynamically generated assets inherit from
- [x] plumb additional asset arguments to the `ExecutionEngine`
- [ ] ~evaluate string types to real python types~ - followup
- [ ] put json schemas under source control
- [x] make other asset types available
  - [x] CSVAsset
  - [x] ExcelAsset
  - [x] JSONAsset
  - [x] ParquetAsset
- [x] verify the generated schemas are reasonable

### Definition of Done

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

## Implementation Notes

`pandas` uses string annotation (from __future__ import annotations) for their io method type annotations.
This means at some point we need to evaluate/resolve these string annotations back into python type objects.
We can...
1. let `pydantic` do this as part of normal model creation/definition. This might require an `.update_forward_references()` call.
2. ~"eval" the string annotations ourselves.~ - (not yet, might be done in follow-up)
    - https://docs.python.org/3/howto/annotations.html#manually-un-stringizing-stringized-annotations
    - [This is how `pydantic` deals with this problem](https://github.com/pydantic/pydantic/blob/v1.10.4/pydantic/typing.py#L375-L404)

Kilo59 committed:
extract IO functions

Kilo59 committed:
extract IO functions

Kilo59 committed:
`DataFrameFactoryFn` alias

Kilo59 committed:
`DataFrameFactoryFn` alias

Kilo59 committed:
extract io signatures

Kilo59 committed:
extract io signatures

ghost commented:
<details>
<summary><b>👇 <a href="https://app.codesee.io/r/reviews?pr=6780&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations">Click on the image</a> for a new way to code review</b>
</summary>

  - Make big changes easier — review code in small groups of related files
  - Know where to start — see the whole change at a glance
  - Take a code tour — explore the change with an interactive tour
  - Make comments and review — all fully sync’ed with github

    [Try it now!](https://app.codesee.io/r/reviews?pr=6780&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)
</details>

[![Review these changes using an interactive CodeSee Map](https://s3.us-east-2.amazonaws.com/maps.codesee.io/images/github/great-expectations/great_expectations/6780/cb5e3321/0ee6647c8478aa1c4a11da8d6f9221d963e2bc4b.svg)](https://app.codesee.io/r/reviews?pr=6780&src=https%3A%2F%2Fgithub.com%2Fgreat-expectations%2Fgreat_expectations)

#### Legend
![CodeSee Map Legend](https://app.codesee.io/Legend_v2.png)

Kilo59 committed:
demo

Kilo59 committed:
demo

Kilo59 committed:
explore param types

Kilo59 committed:
explore param types

Kilo59 committed:
getting `_FieldSpec`s

Kilo59 committed:
getting `_FieldSpec`s

Kilo59 committed:
fix no default check

Kilo59 committed:
fix no default check

Kilo59 committed:
eval'ing annotations

Kilo59 committed:
eval'ing annotations

Kilo59 committed:
cleaner debug/logging

Kilo59 committed:
cleaner debug/logging

Kilo59 committed:
comments

Kilo59 committed:
comments

Kilo59 committed:
exclude fields without primitive types (rough implementation)

using simple string matching

Kilo59 committed:
exclude fields without primitive types (rough implementation)

using simple string matching

Kilo59 committed:
more `CAN_HANDLE` types

Kilo59 committed:
more `CAN_HANDLE` types

Kilo59 committed:
track reason for skipping field

Kilo59 committed:
track reason for skipping field

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
pass down method name

Kilo59 committed:
pass down method name

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
more types and use sets

Kilo59 committed:
more types and use sets

Kilo59 committed:
show all fields that need handling

Kilo59 committed:
show all fields that need handling

Kilo59 committed:
_generate_data_asset_models()

Kilo59 committed:
_generate_data_asset_models()

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
M `TypeVar`

Kilo59 committed:
M `TypeVar`

Kilo59 committed:
Replace CSVAsset with generated version

Kilo59 committed:
Replace CSVAsset with generated version

Kilo59 committed:
working tests

Kilo59 committed:
working tests

Kilo59 committed:
"fix" mypy issues

Kilo59 committed:
"fix" mypy issues

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
`invoke.context.Context` type annotations

Kilo59 committed:
add `str_values()` to `TypeLookup`

Kilo59 committed:
`invoke schema`

Kilo59 committed:
`invoke>=2.0` now allows us to use type annotations without crashing

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
slightly more general `asset_model` creation

Kilo59 committed:
do not register "private" DataAsset classes

Kilo59 committed:
merge

Kilo59 committed:
call `model.update_forward_refs()`

Kilo59 committed:
demo - use fully qualified name instead of injecting locals

example `typing.Sequence`

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
fix asset naming

Kilo59 committed:
`ExcelAsset`, `JSONAsset`, `ParquetAsset`

Kilo59 committed:
better typing for `_get_read_fn()`

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
read from base `pandas` module namespace

Kilo59 committed:
add note and fn annotation

Kilo59 committed:
pass args to `PathBatchSpec`

Kilo59 committed:
`path` doesn't have to be `None`

Kilo59 committed:
sub fields and ignore schema field name error

Kilo59 committed:
pass DataFrameAsset `reader_options`

Kilo59 committed:
`PathBatchSpec` has formal `path` and `reader_options` arguments

Kilo59 committed:
FilePath
`re`

Kilo59 committed:
fix linter

Kilo59 committed:
add `add_<ASSET_TYPE>_asset()` factories for `json`, `excel` & `parquet`

Kilo59 committed:
cleanup

Kilo59 committed:
pandas_csv_data

Kilo59 committed:
`pandas_json_data`

Kilo59 committed:
datasources tasks

Kilo59 committed:
asset json dir

Kilo59 committed:
asset json dir

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
`JSONAsset` `path_or_buff` -> `path`

Kilo59 committed:
working integration test

Kilo59 committed:
passing pandas json asset integration tests

Kilo59 committed:
use SAMPLE_CSV_DIR

Kilo59 committed:
use `pathlib`

Kilo59 committed:
Add `TestDynamicPandasAssets` tests

Kilo59 committed:
comment `path` arg source

Kilo59 committed:
enumerate the pandas io methods

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
lint fixes

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
fix bad merge

Kilo59 commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -246,7 +302,7 @@ def test_pandas_sorter(
     class TimeRange:
         key: str
-        range: List[str]
+        range: list[int]
```
```suggestion
        range: list[str]
```

Kilo59 commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -1,8 +1,13 @@
+import inspect
 from dataclasses import dataclass
-from typing import List
+from pathlib import Path
```
```suggestion
import pathlib
```

Kilo59 committed:
fix bad merge

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +226,96 @@ def add_csv_asset(
             data_path: Path to directory with csv files
             regex: regex pattern that matches csv filenames that is used to label the batches
-            order_by: one of "asc" (ascending) or "desc" (descending) -- the method by which to sort "Asset" parts.
+            kwargs: Extra keyword arguments should correspond to ``pandas.read_csv`` keyword args
```
```suggestion
            order_by: one of "asc" (ascending) or "desc" (descending) -- the method by which to sort "Asset" parts.
            kwargs: Extra keyword arguments should correspond to ``pandas.read_csv`` keyword args
```

Kilo59 commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -148,19 +216,14 @@ def test_get_batch_list_from_fully_specified_batch_request(
 @pytest.mark.unit
 def test_get_batch_list_from_partially_specified_batch_request(
-    pandas_datasource: PandasDatasource, csv_path: pathlib.Path
+    pandas_datasource, csv_path: Path
```
```suggestion
    pandas_datasource: PandasDatasource, csv_path: pathlib.Path
```

Kilo59 committed:
revert changes

Kilo59 committed:
capture fixture

Kilo59 committed:
no really capture it

Kilo59 committed:
working spy & test

Kilo59 committed:
TDD parametrized kwargs

Kilo59 committed:
cleanup

Kilo59 committed:
parse

Kilo59 committed:
more pandas types

Kilo59 committed:
fields without annotations treat as Any

Kilo59 committed:
skiprows

Kilo59 committed:
cleanup

Kilo59 committed:
more types

Kilo59 committed:
cleanup

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
cleanup

Kilo59 committed:
annotations

Kilo59 committed:
use whitelist

Kilo59 committed:
revert test changes

Kilo59 commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -31,6 +52,133 @@ def csv_path() -> pathlib.Path:
+        raising=True,
+    )
+
+    yield captured_args, captured_kwargs
```
This is probably the most important part of these new tests.

It is currently only used by `test_data_asset_reader_options_passthrough`

Kilo59 commented on tests/experimental/datasources/tasks.py:
```diff
@@ -0,0 +1,36 @@
+TEST_ROOT = pathlib.Path(__file__).parent.parent.parent.resolve(strict=True)
+
+
+@invoke.task(aliases=["gen-assets"])
+def generate_asset_files(
+    ctx: invoke.Context,
+    source_csv_dir: str | pathlib.Path = None,
+    limit: int = 12,
+    year: str = 2019,
+):
+    """Generate or re-generate various test asset files."""
```
This is how I generated all the new json files.
We really shouldn't need to run it again unless we want tweak it to generate other pandas formats.

The other thing we could do is just generate these other formats on the fly with some parametrized fixture.
But I didn't want to spend more time going down that rabbit hole.

We discussed before that there's limited value testing all of these various formats because then we are essentially just testing that `pandas` works. 😆 

So TLDR... 
We could do this differently, or not at all.

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
comments

Kilo59 commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -31,6 +52,133 @@ def csv_path() -> pathlib.Path:
+    yield captured_args, captured_kwargs
+
+
+@pytest.mark.unit
+class TestDynamicPandasAssets:
```
I could also put this in it's own test module

Kilo59 commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -177,14 +189,35 @@ def get_batch_list_from_batch_request(
+    ),
+)
+
+CSVAsset = _ASSET_MODELS["csv"]
+ExcelAsset = _ASSET_MODELS["excel"]
+JSONAsset = _ASSET_MODELS["json"]
+ParquetAsset = _ASSET_MODELS["parquet"]
```
We no longer have a model with nicely annotated fields to look at.
One thing we could do is commit the generated json schemas to source control. That way we can catch if they deviate in unexpected ways.

You can use `invoke schema csv|excel|json` etc. to view them. But it's hard to see how things are changing.
Also, it would be trivial to write a tests to make sure the schemas under source control are always in sync with the latest model.

billdirks replied:
This would be good to do.

Kilo59 committed:
fix errors

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
oops

Kilo59 committed:
fix types

Kilo59 committed:
pandas exec engine

Kilo59 committed:
not considered types??

Kilo59 committed:
remove unused type ignores

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
         regex: Union[str, re.Pattern],
         order_by: Optional[BatchSortersDefinition] = None,
-    ) -> CSVAsset:
+        **kwargs,  # TODO: update signature to have specific keys & types
+    ) -> CSVAsset:  # type: ignore[valid-type]
```
I have a follow-up ticket for getting rid of the kwargs

Kilo59 committed:
test debugging

Kilo59 commented:
Note there are some failing integration tests I need to take a look at.
All the existing and new experimental tests pass.

Kilo59 commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -31,6 +52,133 @@ def csv_path() -> pathlib.Path:
+        captured_kwargs.append(kwargs)
+        raise SpyInterrupt(
+            "Reader options have been captured", {"args": args, "kwargs": kwargs}
+        )
```
I actually don't need to do `SpyInterrupt` this capture AND use the mutable `captured_kwargs` containers.
I could do one or the other.

billdirks replied:
Yeah, it seems like you should just put them in `SpyInterrupt` since you are raising it anyways and delete `captured_args` and `captured_kwargs`. The other way is fine too, I'd pick one.

Kilo59 commented on great_expectations/execution_engine/pandas_execution_engine.py:
```diff
@@ -318,6 +323,7 @@ def get_batch_data_and_markers(  # noqa: C901 - 22
+        # ZEP code path
         elif isinstance(batch_spec, PathBatchSpec):
             reader_method = batch_spec.reader_method
             reader_options = batch_spec.reader_options
```
Lines `331` and `332` below are where I inject/monkeypatch my Spy for capturing the given `**reader_options`

alexsherstinsky commented:
A thought regarding this design principle:
```
The goal in this PR is to be able to dynamically create pandas DataAsset types by inspecting the function signatures of the pandas.io read methods (eg read_csv, read_excel, read_json etc.). 20 methods in total.
```

While previously, the burden was on the user to specify the correct reader method and reader options, this approach shifts the burden on us to keep up with the evolution of `Pandas` and other backends' versions and adjust our implementation accordingly as well as manage previous versions, deprecation, and end-of-life scenarios.  This will entail an appreciable increase in support and maintenance.  As long as this approach and costs have been incorporated into our business model, it is in and of itself an acceptable direction, but we should be fully cognizant of the trade-off we are making here.  Thank you.  /cc @donaldheppner @5u2ie

alexsherstinsky commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,284 @@
+import inspect
```
@Kilo59 Since this module does not define a `Datasource` (unless I did not notice that class definition), then perhaps we can call it something like `dynamic_pandas_support.py` (or `dynamic_pandas_helper.py`), etc.?  Thanks.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -135,7 +135,19 @@ def get_batch_list_from_batch_request(
             batch_request
         ):
-            batch_spec = PathBatchSpec(path=str(path))
+            batch_spec = PathBatchSpec(
```
@Kilo59 @billdirks It looks like the direction is toward `DataFrameDataAsset` rather than `FilesystemDataAsset` -- in other words, the focus on the computation (e.g., `DataFrame`), not the source of data (e.g., local filesystem, or cloud filesystem, etc.).  I am curious about the following questions:
* How would we represent an `SQL` data asset?
* How would we represent other `PathBatchSpec` types (e.g., `S3BatchSpec`, `AzureBatchSpec`, `GCSBatchSpec`)?

Thank you.

billdirks replied:
For your SQL question see [TableAsset defined in sql_datasource.py](https://github.com/great-expectations/great_expectations/blob/develop/great_expectations/experimental/datasources/sql_datasource.py#L132).

In the ZEP worldview, the data assets are very much tied to the datasource. @Kilo59 is leveraging that he wants to be very pandas specific so he can leverage the reader method arguments. The `PathBatchSpec` in an implementation detail that he is leveraging so we can use our current execution engine. It is not a ZEP datasource/data asset concept.

So for your 2nd question, do we need an `S3BatchSpec` or will pandas handle this for us if we pass in an `s3` url into `read_csv`? Currently the `S3BatchSpec` implementation is empty. We do switch on the BatchSpec type with `isinstance` checks in some place in code but we should actively try to make those go away and code to a standard interface instead.

I also think it is fine to only support the file system one for now. If we care about the underlying storage layer, we will end up calling a factory here instead of directly instantiating a `PathBatchSpec`.

alexsherstinsky commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,284 @@
+
+    logger.debug(f"Needs extra handling\n{pf(dict(NEED_SPECIAL_HANDLING))}")
+    logger.debug(f"No Annotation\n{FIELD_SKIPPED_NO_ANNOTATION}")
+    return models
```
I would like to see concrete examples of what is being returned that is generalized as ` Type[M]` above.

billdirks replied:
See `pandas_datasource.py` below. These are `CSVAsset`, `ExcelAsset`, etc.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -177,14 +189,35 @@ def get_batch_list_from_batch_request(
         return batch_list
 
 
+_ASSET_MODELS = _generate_data_asset_models(
```
@Kilo59 Per above -- what would be contained in `_ASSET_MODELS` (looking for examples)?  Thanks.

billdirks replied:
These are all subclasses of `DataAsset` defined in `interfaces.py`.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -177,14 +189,35 @@ def get_batch_list_from_batch_request(
+        ExcelAsset,
+        ParquetAsset,
+        JSONAsset,
+    ]
```
@Kilo59 For my edification, are these different assets (which bear file extension names) differ based on the file extension being processed (e.g., ".csv", ".xlsx", ".parquet", ".json")?  Or some other way?  Because if it is by file extension (or, perhaps, also directory-based processing for `Parquet`), then perhaps we can consider the `FilesystemDataAsset` as their parent class (as in #6928 )?  Happy to discuss.  Thanks.

billdirks replied:
These assets are very much tied to pandas while your needs are tied to spark. Maybe we should implement them separately for now and then converge once we have concrete examples of each?

alexsherstinsky replied:
@billdirks Thank you for this -- I think that I am finally starting to have a "Eureka" moment about this.  So, does that mean that we will also need `dynamic_spark.py` -- as a sibling of `dynamic_pandas.py`?

alexsherstinsky replied:
@billdirks @Kilo59 -- What will happen if when executing `PandasDatasource.add_csv_asset()` one supplies the `base_directory_path` (i.e., `data_path: PathStr`) argument and/or the `RegularExpressions` pattern that yield no `CSV` files, but, accidentally, point to a number of files with the ".txt" extension, or ".xlsx" extension, etc. -- i.e., incompatible with the reader methods introspected by the framework for the given `DataAsset` (`CsvDataAsset` in the present example)?  Thank you.

billdirks replied:
We don't validate the file extension. I believe that means the reader method will attempt to read this file and we will get whatever the pandas reader method result is. If the `.txt` really is a `.csv` then we will read it otherwise we'll get a panda's error.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -177,14 +189,35 @@ def get_batch_list_from_batch_request(
     type: Literal["pandas"] = "pandas"
     name: str
-    assets: Dict[str, CSVAsset] = {}
+    assets: Dict[str, Union[CSVAsset, ExcelAsset, ParquetAsset, JSONAsset]] = {}  # type: ignore[valid-type]
```
@Kilo59 This is where `FilesystemDataAsset` (from #6928 ) comes to mind.

billdirks replied:
@Kilo59 I'm not sure how having a `Union` here will interact with the `DataAssetType` `TypeVar` defined in `interfaces.py`. Maybe you should use the parent DataAsset type here or maybe you aren't seeing any type errors from `Datasource.get_asset`?

Kilo59 replied:
Without the type union the schema for `PandasSource` won't contain any of the information about the specific asset subtypes.

But I should say that for our current needs, everything will still work (AFIAK) with `assets: Dict[str, DataAsset]`

Using a `generic` model for our `Datasource` class could help here. But I'd rather do that in a separate PR.
https://docs.pydantic.dev/usage/models/#generic-models

As for whether I get a typing error, I'm going to turn on type-checking for our test code so that we can start seeing these issues more easily.

There is a comment there about the TypeVar needing to change once a Datasource can have more than one asset.

https://github.com/great-expectations/great_expectations/blob/c82b5080d3d3793fd036f3053a8461e5ffea1df0/great_expectations/experimental/datasources/interfaces.py#L292-L293

I don't know if `mpy` considers this union a valid subtype of `DataAsset`, `Union[CSVAsset, ExcelAsset, ParquetAsset, JSONAsset]]` if so then we don't actually need to change the `TypeVar`.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+        )
+        return self.add_asset(asset)
+
+    def add_json_asset(
```
@Kilo59 I believe that the `FilesystemDataAsset` can support JSON as well -- seems no different than CSV, et al.

billdirks replied:
The pandas method `read_csv` and `read_json` have different parameters. If we want to get rid of `kwargs`, display them explicitly in autocomplete (juptyer notebook, for example), and have error messages around them they will need to be different functions on the datasource.
It is true they could all create a `FilesystemDataAsset` in terms of reading them, but @Kilo59 has put validation of these different arguments in the different `DataAsset` subclasses, eg `CSVAsset` validates the input parameters that are CSV specific while `JSONAsset` validates all the arguments that are JSON specific.

Kilo59 replied:
I think some of this might be clearer once I put the schemas under source control.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+            regex: regex pattern that matches csv filenames that is used to label the batches
+            kwargs: Extra keyword arguments should correspond to ``pandas.read_excel`` keyword args
+        """
+        asset = ExcelAsset(
```
@Kilo59 Same observation for Excel.  I could add them all into #6928 -- if we so desire.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+            regex: regex pattern that matches csv filenames that is used to label the batches
+            kwargs: Extra keyword arguments should correspond to ``pandas.read_parquet`` keyword args
+        """
+        asset = ParquetAsset(
```
@Kilo59 We could add `Parquet` to #6928 as well; I am just not 100% certain whether or not the fact that `Parquet` is represented by entire directories "will just work" or if we are supposed to do something special for that.  Thanks.

alexsherstinsky commented on tests/experimental/datasources/tasks.py:
```diff
@@ -0,0 +1,36 @@
+    limit: int = 12,
+    year: str = 2019,
+):
+    """Generate or re-generate various test asset files."""
```
@Kilo59 Are these files automatically cleaned up?  Thanks.

Kilo59 replied:
No this is the utility I used to generate the test JSON files. See my longer explanation below.

https://github.com/great-expectations/great_expectations/pull/6780#discussion_r1087920374

billdirks commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,284 @@
+
+    types: list = []
+
+    union_parts = annotation.split(" | ")
```
This could be fragile. Maybe split on `|` and then call `.strip()` on the parts?

Kilo59 replied:
I agree that's safer, I was trying to avoid extra operations.
My thought was that since pandas is using postponed annotations and `X | Y` the whitespace around the `|` is always going to be consistent.
But I think it makes sense to be safer initially and we can optimize this if needed later.

What I was worried about is that if they forget to use their convention somewhere and we get a `Union[X, Y]` then this will fail.

billdirks commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,284 @@
+        fields = _to_pydantic_fields(signature_tuple)
+
+        type_name = signature_tuple.name.split("read_")[1]
+        model_name = _METHOD_TO_CLASS_NAME_MAPPINGS.get(
```
Can we get rid of this mapping and always do `f"{type_name.capitalize()}Asset"`?

Kilo59 replied:
We can, I was doing this to preserve the uppercase for abbreviations because I thought that was our convention.

@alexsherstinsky asked a similar question.

https://github.com/great-expectations/great_expectations/pull/6780#discussion_r1089017142

billdirks replied:
Oh, I misunderstood this. I thought `.captialize` would capitalize the whole world but it only capitalizes the first letter.

billdirks commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+        )
+        return self.add_asset(asset)
+
+    def add_excel_asset(
```
Are we planning on dynamically generating these `add_*_asset` methods? It seems like we do a lot of work to parse the reader methods and dynamically generate the `DataAsset` classes so we can get them automatically. If we have to manually write the `add_*_asset` methods, that means we still have to do some work to integrate them. 
We definitely don't need to do this (and definitely in this PR) but I was wondering what your thoughts are.

Kilo59 replied:
Yes, I had that same thought.
We had something similar in an earlier version but we decided it was too much magic, without adding a lot of value. 

But as you said now it does dovetail nicely with the dynamic asset creation.

billdirks replied:
Let's file a ticket and do it in a subsequent PR.

billdirks reviewed (commented):
Thanks for doing this @Kilo59! I've left some comments inline. I don't think anything I wrote is blocking. I need to wrap my head around it a bit more. Let's discuss tomorrow.

alexsherstinsky commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,284 @@
+}
+
+_METHOD_TO_CLASS_NAME_MAPPINGS: Final[Dict[str, str]] = {
+    "csv": "CSVAsset",
```
@Kilo59 I would prefer that we spelled out `DataAsset` in the names -- and also formalized the capitalization as in the following examples:
`CsvDataAsset`

`JsonDataAsset`

`HtmlDataAsset`

etc.

There must be published guidelines for capitalization (I have seen some from the ".NET" literature). 

Thanks!

Kilo59 replied:
The naming convention is to keep abbreviations capitalized.
Personally, I think adding `DataAsset` in the name doesn't add much. AFIAK there's no other `Asset` domain object that it could be confused with.

We can change the naming convention if this is what we want, but I don't want to do that in this PR.

billdirks replied:
While I may personally prefer `CsvAsset` over `CSVAsset` we've decided to follow [PEP 8](https://peps.python.org/pep-0008/) and capitalize all acronyms. From the PEP: `Note: When using acronyms in CapWords, capitalize all the letters of the acronym. Thus HTTPServerError is better than HttpServerError.`

alexsherstinsky commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,284 @@
+        )
+
+        try:
+            model = _create_pandas_asset_model(
```
@Kilo59 I would prefer that we spelled out `data_asset` in method names (to avoid confusion down the road).  Thank you.

Kilo59 replied:
I've updated the variable names in this method to `asset_model` and `data_asset_models`

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+        )
+        return self.add_asset(asset)
+
+    def add_excel_asset(
```
@Kilo59 So this method would say `add_excel_data_asset()`.  Thanks.

Kilo59 replied:
Resolving a few of these just so it's easier to read the PR.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+        )
+        return self.add_asset(asset)
+
+    def add_parquet_asset(
```
@Kilo59 And this method would say `add_parquet_data_asset()`.  Thanks.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
+        )
+        return self.add_asset(asset)
+
+    def add_json_asset(
```
@Kilo59 And this method would say `add_json_data_asset()`.  Thanks.

alexsherstinsky commented on great_expectations/experimental/datasources/pandas_datasource.py:
```diff
@@ -201,19 +234,97 @@ def add_csv_asset(
         data_path: PathStr,
         regex: Union[str, re.Pattern],
         order_by: Optional[BatchSortersDefinition] = None,
-    ) -> CSVAsset:
```
@Kilo59 I would prefer that this method would be called `add_csv_data_asset()` -- spelled out for clarity.  Thanks.

Kilo59 replied:
I don't want to change anything about the public API in this PR.

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
safer union splitting

Kilo59 committed:
more descriptive variable names

Kilo59 committed:
test for CLI errors

Kilo59 committed:
revert and add model fallback

Kilo59 committed:
slow replace logic

Kilo59 committed:
don't hide error

Kilo59 commented on great_expectations/experimental/datasources/dynamic_pandas.py:
```diff
@@ -0,0 +1,299 @@
+    default_value: object  # ... for required value
+
+
+@functools.lru_cache(maxsize=64)
+def _replace_builtins(input_: str | type) -> str | type:
+    if not isinstance(input_, str):
+        return input_
+    return input_.replace("list", "List").replace("dict", "Dict")
```
This would be faster as complied regex replace.
Was just trying to prove that this was the issue.

Kilo59 committed:
simplify spy

Kilo59 committed:
remove type alias

billdirks commented on great_expectations/execution_engine/pandas_execution_engine.py:
```diff
@@ -321,6 +326,7 @@ def get_batch_data_and_markers(  # noqa: C901 - 22
             buf.seek(0)
             df = reader_fn(buf, **reader_options)
 
+        # ZEP code path
```
nit: I'd update this comment to make it clear it's not only ZEP that goes down this codepath:
`Experimental datasources will go down this code path`.

Kilo59 replied:
Will do before merge.

Kilo59 committed:
invalid package name

Kilo59 committed:
type-check experimental packages

Kilo59 committed:
ignore everything but `test_viral_snippets`

Kilo59 commented:
~I renamed the `getting-started` test package because it's not a valid python package name and this was causing mypy to crash.~

https://github.com/great-expectations/great_expectations/pull/6780/files#diff-94ecb0e92ccc3d947db32df87bbcc3df310723e87b9f99774dfa6e052e0936f4

Reverted this

Kilo59 committed:
ignore

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
Merge remote-tracking branch 'origin/f/great-1505/zep-dynamic-pandas-assets' into f/great-1505/zep-dynamic-pandas-assets

billdirks commented on tests/experimental/datasources/test_pandas_datasource.py:
```diff
@@ -1,20 +1,41 @@
+logger = logging.getLogger(__file__)
+
+# TODO: use this to parametrize `test_data_asset_defined_for_io_read_method`
+PANDAS_IO_METHODS: list[str] = [
```
Is this used anywhere, can we delete it?

Kilo59 replied:
Will do

billdirks reviewed (approved):
It would be great if you could followup on the Union TypeVar thing we discussed. Otherwise LGTM, thank you!

Kilo59 committed:
cleanup

Kilo59 committed:
better comment

Kilo59 committed:
fix TypeVar usage

Kilo59 committed:
Merge branch 'develop' into f/great-1505/zep-dynamic-pandas-assets

Kilo59 committed:
revert rename

Kilo59 committed:
remove unused

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets

github-actions[bot] committed:
Merge develop into f/great-1505/zep-dynamic-pandas-assets
